[ESIMD] Added test to verify automatic conversion to stateless memory…

v-klochkov opened the pull request:
… access

The option -fsycl-esimd-force-stateless-mem automatically converts accessor
based intrinsics to their versions accepting pointers. This test verifies
gather, scatter, gather_rgba, scatter_rgba, scalar_load, scalar_store,
copy_from, copy_to.

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

kbobrovs commented on SYCL/ESIMD/force_stateless_mem.cpp:
```diff
@@ -0,0 +1,422 @@
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu
+// UNSUPPORTED: cuda || hip
+// RUN: %clangxx -fsycl -fsycl-esimd-force-stateless-mem %s -o %t.out
```
For better coverage, my suggestion would be adding compilation with `-fsycl-esimd-force-stateless-mem` to all existing accessor-based tests. I.e. all of them would contain:
```
// RUN: %clangxx -fsycl %s -o %t.out
// RUN: %clangxx -fsycl -fsycl-esimd-force-stateless-mem %s -o %t_stateless.out
...
// RUN: %GPU_RUN_PLACEHOLDER %t.out
// RUN: %GPU_RUN_PLACEHOLDER %t_stateless.out
```
The same could be done for this test.
Does this make sense?

v-klochkov replied:
That would be less work for me and more testing coverage.
I did not do that because:
- The current tests already take longer time then recommended. (and compilation takes the most of test time in CI)
- Fail of such combined test due to stateless would cause stop of testing of non-stateless functionality as well.

v-klochkov replied:
@kbobrovs - thank you for the comment. I re-worked the PR. Now it splits existing tests using accessor-based API to pairs of tests: the standard-accessor-based and accessor-based-converted-to-stateless.
That is done to avoid having tests in CI that compile/run for too long.
Please review the tests.

v-klochkov committed:
[ESIMD] Add LIT tests for -fsycl-esimd-force-stateless-mem

This option automatically converts accessor-based memory accesses to
stateless accesses.

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

v-klochkov committed:
Fixed errors in lsc tests

Signed-off-by: Vyacheslav N Klochkov <vyacheslav.n.klochkov@intel.com>

kbobrovs reviewed (approved):
Thanks Slava.

v-klochkov commented:
The only failed test is SYCL/Basic/enqueue_barrier.cpp. It is unrelated to this PR and we already have an internal tracker for it.
Merging now.
